gradle test fails for bare React Native app with expo after upgrade to v49

### Minimal reproducible example

https://github.com/TimPoljansek/expo-gradle-test-issue

### Summary

We have a bare React Native app with expo installed. As we have native android modules we run the android tests by executing `./gradlew test` inside the android folder.

But after upgrading to expo 49 running `./gradlew test` no longer works.

This also happens with a fresh bare React Native app after installing expo.

Steps to reproduce:
1. Create a new react native app `npx react-native init example` and install expo `npx install-expo-modules@latest` or download the minimal reproducible example
2. Run `npm install` or `yarn install`
3. Run `cd android`
4. Run `./gradlew test`

Error message:

```
...
> Task :expo-constants:copyReleaseExpoConfig FAILED
...
FAILURE: Build failed with an exception.

* What went wrong:
A problem was found with the configuration of task ':expo-constants:copyReleaseExpoConfig' (type 'Copy').
  - Gradle detected a problem with the following location: '[PROJECT_DIR]/expo-gradle-test-issue/node_modules/expo-constants/android/src/main/assets'.

    Reason: Task ':expo-constants:packageDebugAssets' uses this output of task ':expo-constants:copyReleaseExpoConfig' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed.

    Possible solutions:
      1. Declare task ':expo-constants:copyReleaseExpoConfig' as an input of ':expo-constants:packageDebugAssets'.
      5. Declare an explicit dependency on ':expo-constants:copyReleaseExpoConfig' from ':expo-constants:packageDebugAssets' using Task#dependsOn.
      6. Declare an explicit dependency on ':expo-constants:copyReleaseExpoConfig' from ':expo-constants:packageDebugAssets' using Task#mustRunAfter.

    Please refer to https://docs.gradle.org/8.0.1/userguide/validation_problems.html#implicit_dependency for more details about this problem.

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 51s
200 actionable tasks: 200 executed
``` 

### Environment

  expo-env-info 1.0.5 environment info:
    System:
      OS: macOS 13.4.1
      Shell: 5.9 - /bin/zsh
    Binaries:
      Node: 18.16.1 - ~/.nvm/versions/node/v18.16.1/bin/node
      Yarn: 1.22.19 - /opt/homebrew/bin/yarn
      npm: 9.5.1 - ~/.nvm/versions/node/v18.16.1/bin/npm
      Watchman: 2023.07.10.00 - /opt/homebrew/bin/watchman
    Managers:
      CocoaPods: 1.12.1 - /Users/[USERNAME]/.rbenv/shims/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 22.4, iOS 16.4, macOS 13.3, tvOS 16.4, watchOS 9.4
      Android SDK:
        API Levels: 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33
        Build Tools: 26.0.2, 26.0.3, 27.0.3, 28.0.3, 29.0.0, 29.0.2, 30.0.2, 30.0.3, 31.0.0, 33.0.0, 33.0.1
        System Images: android-22 | Google APIs ARM 64 v8a, android-26 | Google APIs ARM 64 v8a, android-31 | Google APIs ARM 64 v8a, android-33 | Google APIs ARM 64 v8a, android-33 | Google Play ARM 64 v8a
    IDEs:
      Android Studio: 2022.1 AI-221.6008.13.2211.9619390
      Xcode: 14.3.1/14E300c - /usr/bin/xcodebuild
    npmPackages:
      expo: ^49.0.0 => 49.0.2
      react: 18.2.0 => 18.2.0
      react-native: 0.72.1 => 0.72.1
    Expo Workflow: bare

--- a/packages/expo-constants/CHANGELOG.md
+++ b/packages/expo-constants/CHANGELOG.md
@@ -8,7 +8,7 @@
 
 ### 🐛 Bug fixes
 
-- Fix task ':expo-constants:packageDebugAssets' uses this output of task ':expo-constants:copyReleaseExpoConfig' without declaring an explicit or implicit dependency when running `gradlew test` on Android.
+- Fix task ':expo-constants:packageDebugAssets' uses this output of task ':expo-constants:copyReleaseExpoConfig' without declaring an explicit or implicit dependency when running `gradlew test` on Android. ([#23511](https://github.com/expo/expo/pull/23511) by [@lukmccall](https://github.com/lukmccall))
 
 ### 💡 Others
 

--- a/packages/expo-constants/CHANGELOG.md
+++ b/packages/expo-constants/CHANGELOG.md
@@ -8,6 +8,8 @@
 
 ### 🐛 Bug fixes
 
+- Fix task ':expo-constants:packageDebugAssets' uses this output of task ':expo-constants:copyReleaseExpoConfig' without declaring an explicit or implicit dependency when running `gradlew test` on Android.
+
 ### 💡 Others
 
 - Fork `uuid@3.4.0` and move into `expo-modules-core`. Remove the original dependency. ([#23249](https://github.com/expo/expo/pull/23249) by [@alanhughes](https://github.com/alanjhughes))

--- a/packages/expo-constants/scripts/get-app-config-android.gradle
+++ b/packages/expo-constants/scripts/get-app-config-android.gradle
@@ -1,4 +1,4 @@
-// Gradle script for generating serialized public app config (from app.config.js/ts or app.json) and bundling it into the APK 
+// Gradle script for generating serialized public app config (from app.config.js/ts or app.json) and bundling it into the APK
 
 import org.apache.tools.ant.taskdefs.condition.Os
 
@@ -47,20 +47,24 @@ afterEvaluate {
       description = "expo-constants: Copy generated app.config into ${targetName}."
 
       from(assetsDir)
-      into("${projectDir}/src/main/assets")
+      into("${projectDir}/src/${targetPath}/assets")
       include('app.config')
 
-      dependsOn("create${targetName}ExpoConfig")
+      dependsOn(currentBundleTask)
     }
 
     def currentCleanupAppConfigTask = tasks.register("cleanup${targetName}ExpoConfig", Delete) {
       description = "expo-constants: Cleanup generated app.config from ${targetName}."
 
+      // In the past, the `app.config` file was placed in the `main` directory.
+      // However, this is no longer the case. To ensure a smooth transition,
+      // we’re also cleaning the `app.config` from the old location.
       delete files("${projectDir}/src/main/assets/app.config")
+      delete files("${projectDir}/src/${targetPath}/assets/app.config")
       dependsOn(packageTaskName)
     }
 
-    // In summary, these tasks try to create `src/main/assets/app.config` in building time for gradle to merge the asset.
+    // In summary, these tasks try to create `src/${targetPath}/assets/app.config` in building time for gradle to merge the asset.
     // And cleanup the generated app.config after building time.
     tasks.findByPath(packageTaskName).dependsOn(currentCopyAppConfigTask)
     tasks.findByPath("process${targetName}JavaRes").dependsOn(currentCleanupAppConfigTask)

